src/Features/Core/Portable/AddConstructorParametersFromMembers/AddConstructorParametersFromMembersCodeRefactoringProvider.State.cs

```diff
@@ -1,109 +1,140 @@
     {
         private class State
         {
-            public IMethodSymbol ConstructorToAddTo { get; private set; }
+            public ImmutableArray<ConstructorCandidate> ConstructorCandidates { get; private set; }
             public INamedTypeSymbol ContainingType { get; private set; }
-            public ImmutableArray<ISymbol> MissingMembers { get; private set; }
-            public ImmutableArray<IParameterSymbol> MissingParameters { get; private set; }
 
-            public static State Generate(
+            public static async Task<State> GenerateAsync(
                 AddConstructorParametersFromMembersCodeRefactoringProvider service,
-                ImmutableArray<ISymbol> selectedMembers)
+                ImmutableArray<ISymbol> selectedMembers,
+                Document document)
             {
                 var state = new State();
-                if (!state.TryInitialize(service, selectedMembers))
+                if (!await state.TryInitializeAsync(service, selectedMembers, document).ConfigureAwait(true))
                 {
                     return null;
                 }
 
                 return state;
             }
 
-            private bool TryInitialize(
+            private async Task<bool> TryInitializeAsync(
                 AddConstructorParametersFromMembersCodeRefactoringProvider service,
-                ImmutableArray<ISymbol> selectedMembers)
+                ImmutableArray<ISymbol> selectedMembers,
+                Document document)
             {
                 if (!selectedMembers.All(IsWritableInstanceFieldOrProperty))
                 {
                     return false;
                 }
 
-                this.ContainingType = selectedMembers[0].ContainingType;
-                if (this.ContainingType == null || this.ContainingType.TypeKind == TypeKind.Interface)
+                ContainingType = selectedMembers[0].ContainingType;
+                if (ContainingType == null || ContainingType.TypeKind == TypeKind.Interface)
                 {
                     return false;
                 }
 
-                var parameters = service.DetermineParameters(selectedMembers);
+                var parametersForSelectedMembers = service.DetermineParameters(selectedMembers);
                 // We are trying to add these parameters into an existing constructor's parameter list.
                 // Comparing parameters based on names to make sure parameter list won't contains duplicate parameters after we
                 // append the new parameters
-                this.ConstructorToAddTo = GetDelegatedConstructorBasedOnParameterNames(this.ContainingType, parameters);
+                ConstructorCandidates = await GetConstructorCandidatesInfo(ContainingType, parametersForSelectedMembers, selectedMembers, document).ConfigureAwait(false);
 
-                if (this.ConstructorToAddTo == null)
+                if (ConstructorCandidates.Count<ConstructorCandidate>() == 0)
```

Comment:
ImmutableArray<T> has an IsEmpty property
```suggestion
                if (ConstructorCandidates.IsEmpty)
```